File: src/core/export.ts
import chalk from "chalk";
import * as fs from "fs";
import type { Args } from "../cli/args.js";
import { OutputFormat } from "../cli/args.js";
import { formatSize } from "../utils/strings.js";
import type { FileDetail, Summary } from "./types.js";

function displayDirectory(args: Args): string {
  const dir = args.directory;
  return dir === "." ? "current" : dir;
}

function buildHumanReport(summary: Summary, args: Args): string {
  let out = "";

  if (args.quiet) {
    out += `${summary.total_files} ${summary.total_lines}\n`;
    return out;
  }

  out += "=".repeat(60) + "\n";
  out += "LocIO RESULTS\n";
  out += "=".repeat(60) + "\n\n";

  out += `Directory: ${displayDirectory(args)}\n`;

  if (!args.lines_only) {
    out += `\nTotal Files: ${summary.total_files}\n`;
    out += `\nTotal Files Size: ${formatSize(summary.total_size)}\n`;
  }

  if (!args.files_only) {
    out += `\nTotal Lines: ${summary.total_lines}\n`;
  }

  if (Object.keys(summary.files_by_extension).length > 0) {
    const extensions = Object.keys(summary.files_by_extension).sort();
    out += `\nExtensions: ${extensions.join(", ")}\n`;
  }

  if (args.show_stats && Object.keys(summary.files_by_extension).length > 0) {
    const extensions = Object.keys(summary.files_by_extension).sort();

    out += "\nStatistics by Extension:\n";
    out += "-".repeat(60) + "\n";

    for (const ext of extensions) {
      const files = summary.files_by_extension[ext];
      const size = summary.size_by_extension[ext] || 0;
      const lines = summary.lines_by_extension[ext] || 0;

      out += `  ${ext}: ${files} files`;

      if (!args.lines_only) {
        out += `, ${formatSize(size)}`;
      }

      if (!args.files_only) {
        out += `, ${lines} lines`;
      }

      out += "\n";
    }
  }

  if (args.show_stats && summary.details.length > 0) {
    out += "\nFiles by Directory:\n";
    out += "-".repeat(60) + "\n";

    const byDir: Record<string, FileDetail[]> = {};
    for (const detail of summary.details) {
      if (!byDir[detail.directory]) {
        byDir[detail.directory] = [];
      }
      byDir[detail.directory].push(detail);
    }

    const sortedDirs = Object.keys(byDir).sort();
    for (const dir of sortedDirs) {
      const files = byDir[dir];
      out += `Directory: ${dir}\n`;

      for (const f of files) {
        const sizeStr = formatSize(f.size);
        const linesStr =
          f.lines !== null && !args.files_only ? ` | ${f.lines} lines` : "";

        out += `  - ${f.name} (${f.extension}, ${sizeStr}${linesStr})\n`;
      }

      out += "\n";
    }
  }

  out += "\n";

  return out;
}

function humanReport(summary: Summary, args: Args): void {
  if (args.quiet) {
    console.log(`${summary.total_files} ${summary.total_lines}`);
    return;
  }

  console.log("\n" + chalk.cyan("=".repeat(60)));
  console.log(chalk.cyan.bold("LocIO RESULTS"));
  console.log(chalk.cyan("=".repeat(60)));

  console.log(`\n${chalk.green.bold("Directory:")} ${displayDirectory(args)}`);

  if (!args.lines_only) {
    console.log(
      `\n${chalk.green.bold("Total Files:")} ${chalk.yellow(summary.total_files)}`,
    );
    console.log(
      `\n${chalk.green.bold("Total Files Size:")} ${chalk.white(formatSize(summary.total_size))}`,
    );
  }

  if (!args.files_only) {
    console.log(
      `\n${chalk.green.bold("Total Lines:")} ${chalk.yellow(summary.total_lines)}`,
    );
  }

  if (Object.keys(summary.files_by_extension).length > 0) {
    const extensions = Object.keys(summary.files_by_extension).sort();
    console.log(
      `\n${chalk.green.bold("Extensions:")} ${chalk.white(extensions.join(", "))}`,
    );
  }

  if (args.show_stats && Object.keys(summary.files_by_extension).length > 0) {
    console.log(`\n${chalk.cyan.bold("Statistics by Extension:")}`);
    console.log(chalk.gray("-".repeat(60)));

    const extensions = Object.keys(summary.files_by_extension).sort();

    for (const ext of extensions) {
      const files = summary.files_by_extension[ext];
      const size = summary.size_by_extension[ext] || 0;
      const lines = summary.lines_by_extension[ext] || 0;

      process.stdout.write(
        `  ${chalk.white(ext)}: ${chalk.yellow(files)} files`,
      );

      if (!args.lines_only) {
        process.stdout.write(`, ${chalk.white(formatSize(size))}`);
      }

      if (!args.files_only) {
        process.stdout.write(`, ${chalk.yellow(lines)} lines`);
      }

      console.log();
    }
  }

  if (args.show_stats && summary.details.length > 0) {
    console.log(`\n${chalk.cyan.bold("Files by Directory:")}`);
    console.log(chalk.gray("-".repeat(60)));

    const byDir: Record<string, FileDetail[]> = {};
    for (const detail of summary.details) {
      if (!byDir[detail.directory]) {
        byDir[detail.directory] = [];
      }
      byDir[detail.directory].push(detail);
    }

    const sortedDirs = Object.keys(byDir).sort();
    for (const dir of sortedDirs) {
      const files = byDir[dir];
      console.log(chalk.green.bold(`Directory: ${dir}`));

      for (const f of files) {
        const sizeStr = formatSize(f.size);
        const linesStr =
          f.lines !== null && !args.files_only ? ` | ${f.lines} lines` : "";

        console.log(
          `  - ${chalk.white(f.name)} (${chalk.blue(f.extension)}, ${chalk.white(sizeStr)}${linesStr})`,
        );
      }

      console.log();
    }
  }

  console.log();
}

function buildJsonOutput(summary: Summary, args: Args): string {
  const output: any = {
    directory: displayDirectory(args),
    files: summary.total_files,
    size: summary.total_size,
    size_formatted: formatSize(summary.total_size),
  };

  if (!args.files_only) {
    output.lines = summary.total_lines;
  }

  if (args.show_stats) {
    const stats: Record<string, any> = {};
    for (const ext of Object.keys(summary.files_by_extension)) {
      stats[ext] = {
        files: summary.files_by_extension[ext],
        lines: summary.lines_by_extension[ext] || 0,
        size: summary.size_by_extension[ext] || 0,
      };
    }
    output.by_extension = stats;
  }

  return JSON.stringify(output, null, 2);
}

function buildCsvOutput(summary: Summary, args: Args): string {
  let out = `# Directory,${displayDirectory(args)}\n`;
  out += "Extension,Files,Lines,Size\n";
  for (const ext of Object.keys(summary.files_by_extension)) {
    const count = summary.files_by_extension[ext];
    const lines = summary.lines_by_extension[ext] || 0;
    const size = summary.size_by_extension[ext] || 0;
    out += `${ext},${count},${lines},${size}\n`;
  }
  return out;
}

function buildTsvOutput(summary: Summary, args: Args): string {
  let out = `# Directory\t${displayDirectory(args)}\n`;
  out += "Extension\tFiles\tLines\tSize\n";
  for (const ext of Object.keys(summary.files_by_extension)) {
    const count = summary.files_by_extension[ext];
    const lines = summary.lines_by_extension[ext] || 0;
    const size = summary.size_by_extension[ext] || 0;
    out += `${ext}\t${count}\t${lines}\t${size}\n`;
  }
  return out;
}

function writeReportFile(summary: Summary, args: Args): void {
  const format = args.export || OutputFormat.Human;

  let ext: string;
  let content: string;

  switch (format) {
    case OutputFormat.Human:
      ext = "txt";
      content = buildHumanReport(summary, args);
      break;
    case OutputFormat.Json:
      ext = "json";
      content = buildJsonOutput(summary, args);
      break;
    case OutputFormat.Csv:
      ext = "csv";
      content = buildCsvOutput(summary, args);
      break;
    case OutputFormat.Tsv:
      ext = "tsv";
      content = buildTsvOutput(summary, args);
      break;
  }

  const filename = `LocIO-report.${ext}`;

  try {
    fs.writeFileSync(filename, content, "utf-8");
    if (!args.quiet) {
      console.log(`Report written to ${filename}`);
    }
  } catch (e) {
    console.error(`Failed to create report file ${filename}: ${e}`);
  }
}

export function exportReport(summary: Summary, args: Args): void {
  if (args.export !== undefined) {
    writeReportFile(summary, args);
    return;
  }

  humanReport(summary, args);
}

File: src/export/export.rs
use crate::cli::{Args, OutputFormat};
use crate::core::{Summary, FileDetail};
use crate::utils;
use colored::*;
use serde_json::json;
use std::fs;
use std::io::Write;

fn display_directory(args: &Args) -> String {
    let dir = args.directory.display().to_string();
    if dir == "." {
        "current".to_string()
    } else {
        dir
    }
}

fn build_human_report(summary: &Summary, args: &Args) -> String {
    let mut out = String::new();

    if args.quiet {
        out.push_str(&format!("{} {}\n", summary.total_files, summary.total_lines));
        return out;
    }

    out.push_str(&format!("{}\n", "=".repeat(60)));
    out.push_str("LINE COUNTER RESULTS\n");
    out.push_str(&format!("{}\n\n", "=".repeat(60)));

    out.push_str(&format!("Directory: {}\n", display_directory(args)));

    if !args.lines_only {
        out.push_str(&format!(
            "\nTotal Files: {} ({})\n",
            summary.total_files,
            utils::format_size(summary.total_size)
        ));
    }

    if !args.files_only {
        out.push_str(&format!("\nTotal Lines: {}\n", summary.total_lines));
    }

    if !summary.files_by_extension.is_empty() {
        let mut extensions: Vec<_> = summary.files_by_extension.keys().collect();
        extensions.sort();
        let ext_list: Vec<String> = extensions.iter().map(|ext| ext.to_string()).collect();
        out.push_str(&format!("\nExtensions: {}\n", ext_list.join(", ")));
    }

    if args.show_stats && !summary.files_by_extension.is_empty() {
        let mut extensions: Vec<_> = summary.files_by_extension.keys().collect();
        extensions.sort();

        out.push_str("\nStatistics by Extension:\n");
        out.push_str(&format!("{}\n", "-".repeat(60)));

        for ext in extensions {
            let files = summary.files_by_extension[ext];
            let size = summary.size_by_extension.get(ext).copied().unwrap_or(0);
            let lines = summary.lines_by_extension.get(ext).copied().unwrap_or(0);

            out.push_str(&format!("  {}: {} files", ext, files));

            if !args.lines_only {
                out.push_str(&format!(", {}", utils::format_size(size)));
            }

            if !args.files_only {
                out.push_str(&format!(", {} lines", lines));
            }

            out.push('\n');
        }
    }

    if args.show_stats && !summary.details.is_empty() {
        use std::collections::BTreeMap;

        out.push_str("\nFiles by Directory:\n");
        out.push_str(&format!("{}\n", "-".repeat(60)));

        let mut by_dir: BTreeMap<&str, Vec<&FileDetail>> = BTreeMap::new();
        for detail in &summary.details {
            by_dir
                .entry(detail.directory.as_str())
                .or_default()
                .push(detail);
        }

        for (dir, files) in by_dir {
            out.push_str(&format!("Directory: {}\n", dir));

            for f in files {
                let size_str = utils::format_size(f.size);
                let lines_str = match f.lines {
                    Some(l) if !args.files_only => format!(" | {} lines", l),
                    _ => String::new(),
                };

                out.push_str(&format!(
                    "  - {} ({}, {}{})\n",
                    f.name, f.extension, size_str, lines_str
                ));
            }

            out.push('\n');
        }
    }

    out.push('\n');

    out
}

pub fn human_report(summary: &Summary, args: &Args) {
    if args.quiet {
        println!("{} {}", summary.total_files, summary.total_lines);
        return;
    }

    println!("\n{}", "=".repeat(60).bright_cyan());
    println!("{}", "LINE COUNTER RESULTS".bright_cyan().bold());
    println!("{}", "=".repeat(60).bright_cyan());

    println!(
        "\n{} {}",
        "Directory:".bright_green().bold(),
        display_directory(args)
    );

    if !args.lines_only {
        println!(
            "\n{} {} {}",
            "Total Files:".bright_green().bold(),
            summary.total_files.to_string().bright_yellow(),
            utils::format_size(summary.total_size).bright_white()
        );
    }

    if !args.files_only {
        println!(
            "\n{} {}",
            "Total Lines:".bright_green().bold(),
            summary.total_lines.to_string().bright_yellow()
        );
    }

    if !summary.files_by_extension.is_empty() {
        let mut extensions: Vec<_> = summary.files_by_extension.keys().collect();
        extensions.sort();

        print!("\n{} ", "Extensions:".bright_green().bold());
        let ext_list: Vec<String> = extensions.iter().map(|ext| ext.to_string()).collect();
        println!("{}", ext_list.join(", ").bright_white());
    }

    if args.show_stats && !summary.files_by_extension.is_empty() {
        println!("\n{}", "Statistics by Extension:".bright_cyan().bold());
        println!("{}", "-".repeat(60).bright_black());

        let mut extensions: Vec<_> = summary.files_by_extension.keys().collect();
        extensions.sort();

        for ext in extensions {
            let files = summary.files_by_extension[ext];
            let size = summary.size_by_extension.get(ext).copied().unwrap_or(0);
            let lines = summary.lines_by_extension.get(ext).copied().unwrap_or(0);

            print!(
                "  {}: {} files",
                ext.bright_white(),
                files.to_string().bright_yellow()
            );

            if !args.lines_only {
                print!(", {}", utils::format_size(size).bright_white());
            }

            if !args.files_only {
                print!(", {} lines", lines.to_string().bright_yellow());
            }

            println!();
        }
    }

    if args.show_stats && !summary.details.is_empty() {
        use std::collections::BTreeMap;

        println!("\n{}", "Files by Directory:".bright_cyan().bold());
        println!("{}", "-".repeat(60).bright_black());

        let mut by_dir: BTreeMap<&str, Vec<&FileDetail>> = BTreeMap::new();
        for detail in &summary.details {
            by_dir
                .entry(detail.directory.as_str())
                .or_default()
                .push(detail);
        }

        for (dir, files) in by_dir {
            println!(
                "{}",
                format!("Directory: {}", dir).bright_green().bold()
            );

            for f in files {
                let size_str = utils::format_size(f.size);
                let lines_str = match f.lines {
                    Some(l) if !args.files_only => format!(" | {} lines", l),
                    _ => String::new(),
                };

                println!(
                    "  - {} ({}, {}{})",
                    f.name.bright_white(),
                    f.extension.bright_blue(),
                    size_str.bright_white(),
                    lines_str
                );
            }

            println!();
        }
    }

    println!();
}

fn build_json_output(summary: &Summary, args: &Args) -> String {
    let mut output = json!({
        "directory": display_directory(args),
        "files": summary.total_files,
        "size": summary.total_size,
        "size_formatted": utils::format_size(summary.total_size),
    });

    if !args.files_only {
        output["lines"] = json!(summary.total_lines);
    }

    if args.show_stats {
        let mut stats = json!({});
        for (ext, count) in &summary.files_by_extension {
            stats[ext] = json!({
                "files": count,
                "lines": summary.lines_by_extension.get(ext).copied().unwrap_or(0),
                "size": summary.size_by_extension.get(ext).copied().unwrap_or(0),
            });
        }
        output["by_extension"] = stats;
    }

    serde_json::to_string_pretty(&output).unwrap()
}

fn build_csv_output(summary: &Summary, _args: &Args) -> String {
    let mut out = String::new();
    out.push_str(&format!("# Directory,{}\n", display_directory(_args)));
    out.push_str("Extension,Files,Lines,Size\n");
    for (ext, count) in &summary.files_by_extension {
        let lines = summary.lines_by_extension.get(ext).copied().unwrap_or(0);
        let size = summary.size_by_extension.get(ext).copied().unwrap_or(0);
        out.push_str(&format!("{},{},{},{}\n", ext, count, lines, size));
    }
    out
}

fn build_tsv_output(summary: &Summary, _args: &Args) -> String {
    let mut out = String::new();
    out.push_str(&format!("# Directory\t{}\n", display_directory(_args)));
    out.push_str("Extension\tFiles\tLines\tSize\n");
    for (ext, count) in &summary.files_by_extension {
        let lines = summary.lines_by_extension.get(ext).copied().unwrap_or(0);
        let size = summary.size_by_extension.get(ext).copied().unwrap_or(0);
        out.push_str(&format!("{}\t{}\t{}\t{}\n", ext, count, lines, size));
    }
    out
}

fn write_report_file(summary: &Summary, args: &Args) {
    let format = args.export.clone().unwrap_or(OutputFormat::Human);

    let (ext, content) = match format {
        OutputFormat::Human => ("txt", build_human_report(summary, args)),
        OutputFormat::Json => ("json", build_json_output(summary, args)),
        OutputFormat::Csv => ("csv", build_csv_output(summary, args)),
        OutputFormat::Tsv => ("tsv", build_tsv_output(summary, args)),
    };

    let filename = format!("locio-report.{}", ext);

    match fs::File::create(&filename) {
        Ok(mut file) => {
            if let Err(e) = file.write_all(content.as_bytes()) {
                eprintln!("Failed to write report to {}: {}", filename, e);
            } else if !args.quiet {
                println!("Report written to {}", filename);
            }
        }
        Err(e) => {
            eprintln!("Failed to create report file {}: {}", filename, e);
        }
    }
}

pub fn export_report(summary: &Summary, args: &Args) {
    if args.export.is_some() {
        write_report_file(summary, args);
        return;
    }

    human_report(summary, args);
}
